pyscriptjs/tests/integration/support.py

```diff
@@ -92,6 +93,32 @@ def init(self, request, tmpdir, http_server, logger, page):
 
     def init_page(self, page):
         self.page = page
+        # cache to store data via sha256(url)
+        cache = {}
+
+        # router to cache with fail 2x on requests
+
+        def router(route):
+            # hash of url
+            hash = hashlib.sha256(route.request.url.encode("utf-8")).hexdigest()
+            # cached?
+            if hash in cache:
+                # fulfill via cache
+                route.fulfill(status=200, response=cache.get(hash))
+            else:
+                try:
+                    # fetch, cache, fulfill 1x
+                    response = page.request.fetch(route.request)
+                    cache[hash] = response
+                    route.fulfill(status=200, response=response)
+                except Exception:
+                    # fetch, cache, fulfill 2x
+                    response = page.request.fetch(route.request)
+                    cache[hash] = response
+                    route.fulfill(status=200, response=response)
```

Comment:
In my opinion, the snippets under try and catch are exactly the same, so it makes sense to make a function out of this and put it inside a retry decorator.

So, for example:

```
@retry(times=2, exceptions=(Exception,))
def fetch_and_put_in_cache():
    response = page.request.fetch(route.request)
    cache[hash] = response
    route.fulfill(status=200, response=response)
```

```
def retry(times, exceptions):
    def decorator(func):
        def newfn(*args, **kwargs):
            attempt = 0
            while attempt < times:
                try:
                    return func(*args, **kwargs)
                except exceptions:
                    logging.warning(
                        f"Exception thrown when attempting to run {func} at attempt {attempt} of {times}, retrying..."
                    )
                    attempt += 1
            return func(*args, **kwargs)

        return newfn

    return decorator
```

Reply:
These are at very different levels:
One means the entire test reruns
One means we retry a web request within the test.

Most of the test flakiness is due to attempting to load URL within the test browser and these fail/hang. This also sets up the opportunity to run all integration tests without a web server and run them in parallel.

Reply:
What I mean is that -- a decorator would give us the opportunity to:
- control the number of retries without cascading try...except blocks, so we can simply change number of retries to 5 let's say
- control the exceptions on which we want to retry

further, this decorator could be re-usable to other parts later, right now it fits for your use-case of retry a web request within the test.

but you can also apply it at both levels if needed

Reply:
Decorator for simplifying logic. Thanks @madhur-tandon

Reply:
resolving...